### PR TITLE
Option to configure the ZFS block size on volumes

### LINKED ISSUE #7901: Option to configure the ZFS block size on volumes
Custom volumes on ZFS (both filesystems and zvols) should expose control of their `recordsize` or `volblocksize` properties respectively.
This is commonly used to "tune" (read: fix) performance of applications like databases on ZFS, where it can have a huge impact due to its COW nature.

### PR DESCRIPTION
**Pull request by presztak**
Closes #7901

### CODE REVIEW

File: lxd/storage/drivers/driver_zfs_utils.go [hunk 1]
```diff
@@ -13,10 +13,18 @@ import (
...
 )
 
-// zfsBlockVolSuffix suffix used for block content type volumes.
-const zfsBlockVolSuffix = ".block"
+const (
+	// zfsBlockVolSuffix suffix used for block content type volumes.
+	zfsBlockVolSuffix = ".block"
+
+	// zfsMinBlockSize is a minimum value for recordsize and volblocksize properties.
+	zfsMinBlocksize = 512
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_utils.go:24]: line break after this one

File: lxd/storage/drivers/driver_zfs_utils.go [hunk 2]
```diff
@@ -13,10 +13,18 @@ import (
...
+	// zfsMinBlockSize is a maximum value for recordsize and volblocksize properties.
+	zfsMaxBlocksize = 1024 * 1024
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_utils.go:26]: Sounds like we'll be bumping that one to 16 * 1024 * 1024 then

File: lxd/storage/drivers/driver_zfs_utils.go [hunk 3]
```diff
@@ -367,3 +401,27 @@ func (d *zfs) receiveDataset(vol Volume, conn io.ReadWriteCloser, writeWrapper f
...
+// ValidateZfsBlocksize validates blocksize property value on the pool.
+func ValidateZfsBlocksize(value string) error {
+	// Convert to bytes.
+	sizeBytes, err := units.ParseByteSizeString(value)
+	if err != nil {
+		return err
+	}
+
+	if sizeBytes < zfsMinBlocksize || sizeBytes > zfsMaxBlocksize || (sizeBytes&(sizeBytes-1)) != 0 {
+		return fmt.Errorf("Value should be between 512 and 1MiB, and be power of 2")
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_utils.go:414]: so 16MiB now

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 4]
```diff
@@ -31,6 +31,9 @@ import (
 	"github.com/lxc/lxd/shared/validate"
 )
 
+// zfsMaxVolBlocksize is a maximum value for volblocksize property.
+const zfsMaxVolBlocksize = 128 * 1024
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_volumes.go:35]: I'd keep that const with the others in driver_zfs.go

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 5]
```diff
@@ -158,6 +167,22 @@ func (d *zfs) CreateVolume(vol Volume, filler *VolumeFiller, op *operations.Oper
...
 
+		blockSize := vol.ExpandedConfig("zfs.blocksize")
+		if blockSize != "" {
+			// Convert to bytes.
+			sizeBytes, err := units.ParseByteSizeString(blockSize)
+			if err != nil {
+				return err
+			}
+
+			// zfs.blocksize can have value in range from 512 to 1MiB because it's used for volblocksize and recordsize
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_volumes.go:178]: 16MiB

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 6]
```diff
@@ -158,6 +167,22 @@ func (d *zfs) CreateVolume(vol Volume, filler *VolumeFiller, op *operations.Oper
...
+			// volblocksize maximum value is 128KiB so if the value of zfs.blocksize is bigger set it to 128KiB.
+			if sizeBytes > zfsMaxVolBlocksize {
+				sizeBytes = zfsMaxVolBlocksize
+			}
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_volumes.go:182]: line break

File: lxd/storage/drivers/driver_zfs_volumes.go [hunk 7]
```diff
@@ -834,6 +878,17 @@ func (d *zfs) UpdateVolume(vol Volume, changedConfig map[string]string) error {
 			old[k] = vol.config[k]
 			vol.config[k] = v
 		}
+		if k == "zfs.blocksize" {
+			// Convert to bytes.
+			sizeBytes, err := units.ParseByteSizeString(v)
+			if err != nil {
+				return err
+			}
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_volumes.go:886]: line break

File: lxd/storage/drivers/driver_zfs_utils.go [hunk 8]
```diff
@@ -367,3 +401,27 @@ func (d *zfs) receiveDataset(vol Volume, conn io.ReadWriteCloser, writeWrapper f
...
+	}
+	return nil
+}
+
+// ValidateZfsVolBlocksize validates blocksize property value on the volume.
+func ValidateZfsVolBlocksize(vol Volume) func(value string) error {
+	return func(value string) error {
+		if vol.contentType != ContentTypeFS {
+			return fmt.Errorf("Blocksize can be change only for filesystem type")
+		}
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_utils.go:424]: line break

File: lxd/storage/drivers/driver_zfs_utils.go [hunk 9]
```diff
@@ -195,6 +203,32 @@ func (d *zfs) setDatasetProperties(dataset string, options ...string) error {
...
+
+func (d *zfs) setBlocksize(vol Volume, size int64) error {
+	if vol.contentType != ContentTypeFS {
+		return nil
+	}
+
+	err := d.setDatasetProperties(d.dataset(vol, false), fmt.Sprintf("recordsize=%d", size))
+	if err != nil {
+		return err
+	}
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_utils.go:228]: line break

File: lxd/storage/drivers/driver_zfs_utils.go [hunk 10]
```diff
@@ -195,6 +203,32 @@ func (d *zfs) setDatasetProperties(dataset string, options ...string) error {
...
+	size := vol.ExpandedConfig("zfs.blocksize")
+	if size == "" {
+		return nil
+	}
+
+	// Convert to bytes.
+	sizeBytes, err := units.ParseByteSizeString(size)
+	if err != nil {
+		return err
+	}
```
Comment (stgraber) [inline lxd/storage/drivers/driver_zfs_utils.go:216]: line break

**Comment by stgraber** [comment]
Looking great!

Just a few formatting nits, a const to relocate and the bump to 16MiB and we'll be good to go!

File: lxd/storage/drivers/driver_zfs_utils.go (hunk 10 above)
Comment (presztak) [reply lxd/storage/drivers/driver_zfs_utils.go:220]: Fixed

File: lxd/storage/drivers/driver_zfs_utils.go (hunk 9 above)
Comment (presztak) [reply lxd/storage/drivers/driver_zfs_utils.go:233]: Fixed

File: lxd/storage/drivers/driver_zfs_utils.go (hunk 8 above)
Comment (presztak) [reply lxd/storage/drivers/driver_zfs_utils.go:431]: Fixed

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 6 above)
Comment (presztak) [reply lxd/storage/drivers/driver_zfs_volumes.go:179]: Fixed

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 7 above)
Comment (presztak) [reply lxd/storage/drivers/driver_zfs_volumes.go:884]: Fixed

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 5 above)
Comment (presztak) [reply lxd/storage/drivers/driver_zfs_volumes.go]: Fixed

File: lxd/storage/drivers/driver_zfs_volumes.go (hunk 4 above)
Comment (presztak) [reply lxd/storage/drivers/driver_zfs_volumes.go]: Fixed

File: lxd/storage/drivers/driver_zfs_utils.go (hunk 3 above)
Comment (presztak) [reply lxd/storage/drivers/driver_zfs_utils.go]: Fixed

File: lxd/storage/drivers/driver_zfs_utils.go (hunk 2 above)
Comment (presztak) [reply lxd/storage/drivers/driver_zfs_utils.go]: Fixed

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Thanks!

**Pull request merged**